pytest_freezegun.py

```diff
@@ -1,18 +1,48 @@
 # -*- coding: utf-8 -*-
 
 import pytest
+
 from freezegun import freeze_time
 
 
-@pytest.yield_fixture
+@pytest.fixture
 def freezer(request):
-    marker = request.node.get_marker('freeze_time')
-    args, kwargs = (marker.args, marker.kwargs) if marker else ((), {})
-    with freeze_time(*marker.args, **marker.kwargs) as freezer:
-        yield freezer
+    plugin = request.config.pluginmanager.get_plugin('freezer-deferred')
+
+    if plugin.frozen_time is None:
+        with freeze_time() as frozen_time:
+            yield frozen_time
+    else:
+        yield plugin.frozen_time
+
+
+class FreezegunPlugin(object):
+    __name__ = 'freezer-deferred'
+
+    def __init__(self):
+        self.freezer = None
+        self.frozen_time = None
+
+    @pytest.hookimpl(tryfirst=True)
+    def pytest_runtest_setup(self, item):
+        marker = item.get_marker('freeze_time')
+
+        if marker:
+            ignore = marker.kwargs.pop('ignore', [])
+            ignore.append('_pytest')
+
+            self.freezer = freeze_time(
+                *marker.args,
+                ignore=ignore,
+                **marker.kwargs
+            )
+            self.frozen_time = self.freezer.start()
+
+    @pytest.hookimpl(trylast=True)
+    def pytest_runtest_teardown(self):
+        if self.freezer is not None:
+            self.freezer.stop()
 
 
-@pytest.fixture(autouse=True)
-def _auto_freezer(request):
-    if request.node.get_marker('freeze_time'):
-        request.getfixturevalue('freezer')
+def pytest_configure(config):
+    config.pluginmanager.register(FreezegunPlugin())
```

Comment:
Why do you defer the plugin's initialization?

Reply:
I want to store `freezer` and `frozen_time` as instance attributes instead of global variables, so the actual plugin implementation lives in the `FreezegunPlugin` class. As far as I'm aware deferring it is the only way to actually make it work as a plugin, pytest doesn't offer a better API for that.

Technically, those vars could be stored as globals, attributes on `item`, or perhaps somewhere else, but I find . these solutions quite ugly in their nature.

If you have an idea how to approach in a better way, I'm happy to change it :)

Reply:
You could just create a global instance of the plugin and change the entrypoint in setup.py to `pytest_freezegun:plugin` - it'll act as a singleton anyway. Also, you can move the freezer fixture into the plugin class, this way you won't have to touch pluginmanager at all.

I wonder what using a class, as opposed to just a module with globals, actually buys us - it's maybe easier to override behaviour by subclassing?

Reply:
Clearly, I was wrong - it does offer a better API. I have changed it to create the plugin instance and use it as an entry point. Also, moved the fixture into the class.

As to why keep it as a class - I think state should live in an object. Keeping it in global variables is not elegant.